packages/notifier/src/publish-kit.js

```diff
@@ -180,12 +200,221 @@ export const makePublishKit = () => {
+      new Error('Cannot read past end of iteration.'),
+    );
+    newData = {
+      currentP: HandledPromise.resolve(
+        harden({
+          head: { value: state.value, done: true },
+          publishCount,
+          tail: tailP,
+        }),
+      ),
+      tailP,
+      tailR: undefined,
+    };
+  } else if (status === 'live') {
+    const { promise: tailP, resolve: tailR } = makePromiseKit();
+    newData = {
+      currentP: state.hasValue
+        ? HandledPromise.resolve(
+            harden({
+              head: { value: state.value, done: false },
+              publishCount,
+              tail: tailP,
+            }),
+          )
+        : tailP,
+      tailP,
+      tailR,
+    };
+  } else {
+    assert.fail(X`Invalid durable promise kit status: ${q(status)}`);
+  }
+  durablePublishKitEphemeralData.set(state, harden(newData));
+  return newData;
+};
+
+/**
+ * Extends the sequence of results.
+ *
+ * @param {DurablePublishKitState} state
+ * @param {boolean} done
+ * @param {any} value
+ * @param {any} [rejection]
+ */
+const advanceDurablePublishKit = (state, done, value, rejection) => {
+  const { valueDurability, status } = state;
+  if (status !== 'live') {
+    throw new Error('Cannot update state after termination.');
+  }
+  if (done || valueDurability === 'mandatory') {
+    canBeDurable(value) ||
+      assert.fail(X`Cannot accept non-durable value: ${value}`);
+    canBeDurable(rejection) ||
+      assert.fail(X`Cannot accept non-durable rejection: ${rejection}`);
+  }
+  const { tailP: currentP, tailR: resolveCurrent } =
+    provideDurablePublishKitEphemeralData(state);
+
+  state.publishCount += 1n;
+  const publishCount = state.publishCount;
+  let tailP;
+  let tailR;
+
+  if (done) {
+    state.status = rejection ? 'failed' : 'finished';
+    tailP = makeQuietRejection(new Error('Cannot read past end of iteration.'));
+    tailR = undefined;
+  } else {
+    ({ promise: tailP, resolve: tailR } = makePromiseKit());
+  }
+  durablePublishKitEphemeralData.set(state, harden({ currentP, tailP, tailR }));
+
+  if (rejection) {
+    state.value = rejection;
+    state.hasValue = true;
+    resolveCurrent(rejection);
+  } else {
+    state.hasValue = false;
+    try {
+      if (done || valueDurability !== 'ignored') {
+        canBeDurable(value) ||
+          assert.fail(X`Cannot accept non-durable value: ${value}`);
+        state.value = value;
+        state.hasValue = true;
+      }
+    } catch (err) {
+      if (done || valueDurability === 'mandatory') {
+        throw err;
+      }
+    }
+    resolveCurrent(
+      harden({
+        head: { value, done },
+        publishCount,
+        tail: tailP,
+      }),
+    );
+  }
+};
+
+/**
+ * @param {import('../../vat-data/src/types.js').Baggage} baggage
+ * @param {string} kindName
+ */
+export const vivifyDurablePublishKit = (baggage, kindName) => {
+  /**
+   * @returns {() => PublishKit<*>}
+   */
+  return vivifyFarClassKit(
```

Comment:
Still kinda disturbed to have two similar but different implementations. I suspect the answer is no, but it is worth asking:

Would it help to use `defineHeapFarClassKit` for the ephemeral case, so that the ephemeral case and the durable case can share more code?

Reply:
It might be possible, but I don't want to make the attempt until the new implementation is stable and confirmed to survive upgrade.